dmipy/core/fitted_modeling_framework.py

```diff
@@ -897,3 +897,246 @@ def mean_squared_error(self, data):
+        vertices = sphere.vertices
+        _, inv_sh_matrix = sh_to_sf_matrix(
+            sphere, sh_order, basis_type=basis_type, return_inv=True)
+        fods_sf = self.fod(vertices)
+
+        dataset_shape = self.fitted_parameters_vector.shape[:-1]
+        number_coef_used = int((sh_order + 2) * (sh_order + 1) // 2)
+        fods_sh = np.zeros(np.r_[dataset_shape, number_coef_used])
+        mask_pos = np.where(self.mask)
+        for pos in zip(*mask_pos):
+            fods_sh[pos] = np.dot(inv_sh_matrix.T, fods_sf[pos])
+        return fods_sh
+
+    def peaks_spherical(self):
+        "Returns the peak angles of the model."
+        mu_params = []
+        for name, card in self.model.parameter_cardinality.items():
+            if name[-2:] == 'mu' and card == 2:
+                mu_params.append(self.fitted_parameters[name])
+        if len(mu_params) == 0:
+            msg = ('peaks not available for current model.')
+            raise ValueError(msg)
+        if len(mu_params) == 1:
+            return np.expand_dims(mu_params[0], axis=-2)
+        return np.concatenate([mu[..., None] for mu in mu_params], axis=-1)
+
+    def peaks_cartesian(self):
+        "Returns the cartesian peak unit vectors of the model."
+        peaks_spherical = self.peaks_spherical()
+        peaks_cartesian = unitsphere2cart_Nd(peaks_spherical)
+        return peaks_cartesian
+
+    def predict(self, acquisition_scheme=None, S0=None, mask=None):
+        """
+        simulates the dMRI signal of the fitted MultiCompartmentModel for the
+        estimated model parameters. If no acquisition_scheme is given, then
+        the same acquisition_scheme that was used for the fitting is used. If
+        no S0 is given then it is assumed to be the estimated one. If no mask
+        is given then all voxels are assumed to have been fitted.
+
+        Parameters
+        ----------
+        acquisition_scheme : DmipyAcquisitionScheme instance,
+            An acquisition scheme that has been instantiated using dMipy.
+        S0 : None or float,
+            Signal intensity without diffusion sensitization. If None, uses
+            estimated SO from fitting process. If float, uses that value.
+        mask : (N-1)-dimensional integer/boolean array of size (N_x, N_y, ...),
+            mask of voxels to simulate data at.
+
+        Returns
+        -------
+        predicted_signal : array of size (Ndata, N_DWIS),
+            predicted DWIs for the given model parameters and acquisition
+            scheme.
+        """
+        if acquisition_scheme is None:
+            # TODO: for each voxel, multiply the forward model matrix by the
+            #  corresponding x.
+            pass
+        else:
+            # TODO: recompute the forward model matrix on the given set of
+            #  orientations. Also, the coefficients for each orientation must
+            #  be resampled.
+            pass
+        raise NotImplementedError
```

Comment:
I don't think the predict function should be nonimplemented in this first MR since it addresses basic functionality of the fitted model.

Since in the first version we will not even have the lookup table, we will have to calculate the forward model matrix no matter if `acquisition_scheme=None` or something else.

The code you wrote here appropriate for the follow-up MR with look-up table.

Reply:
let's define if the LUT is part of the first MR or not ;)

Reply:
We agreed on _no LUT_ for the moment.

Anyway, we need to think this properly. If we have the parameters fitted on a predefined set of orientations, how do we "register" this to the new acquisition scheme? Do we want to use a spherical harmonics representation and resample it accordingly?

Reply:
Still, it would be tricky for different shells

Reply:
I think the function should look something like

```
if acquisition_scheme is None:
    acquisition_scheme = self.scheme
E_shape = self.model.mask.shape + (acquisition_scheme.N_dwi,)
E = np.zeros(E_shape)
x, y, z = np.where(self.mask)
for x_, y_, z_ in zip(x, y, z):
    parameter_array = self.fitted_parameter_vectors[x_, y_, z_]
    M = self.forward_model(xxxxx)  # same way as we instantiated it before the fitting
    E[x_, y_, z_] = M.dot(parameter_array)
return E
```